### ISSUE TITLE
Feature request: edit by section

### ISSUE THREAD

**Issue by nschneid** [gollum/gollum#715]
For large documents with many sections, making a small edit can require tedious searching on the edit page. Wikipedia-style "edit section" links on the page would improve this experience.

**Comment by dometto** [comment]
Just my $0.02: treating sections as files would require overhauling *a lot* of gollum's internal workings (TOC/anchors, rendering). I think it should be possible to have a relatively clean implementation of per-section editing by just hiding the rest of the document through JS magic.

**Comment by cniedzwiedz** [comment]
I also think this would be a great feature.  I've needed it constantly since using starting with Gollum.  Unfortunately, I don't have a good knowledge of Gollum internals and can't comment on the proposed solutions.

**Comment by dometto** [comment]
Some further thoughts:
* This is much more difficult than I thought, because I see no general way to get the portion of raw markup that corresponds to a html section (the content between two headings).
* We could try to manually parse markup into sections and determine how they correspond to the `id`'s of header-links in the rendered page. While that's still relatively easy for markdown, we'd have to do this for **every** markup type.
    * *Or* we only support by-section editing on markdown

* This applies however we want to actually implement the by-section edits: by 'hiding' the rest of the document, or by treating sections as separate files. In each case, we need a mapping from a rendered page's edit-heading links to a portion of the raw markup document.

TL;DR I'm at a loss about how this could be achieved in gollum. :)

**Comment by dometto** [comment]
FYI, proof-of-concept implementation of markdown section-splitting in ruby via kramdown:
https://gist.github.com/dometto/1becfed8b48bb2799989
Perhaps it's possible to do this with other markup gems, too.

**Comment by dometto** [comment]
Proof-of-concept implementation in gollum:
https://github.com/repotag/gollum-lib/tree/edit_sections
...and:
https://github.com/repotag/gollum/tree/edit_section

Going to `http://localhost:4567/edit/Page?section=anchor-id-of-section', and editing/saving the section, actually works! However, this only happens when `kramdown` is available, and will only actually work for markdown pages.

The section-parser uses the same logic for computing header id's as the TOC filter, so if you have a `#header` anchor on a page, `/page?section=header`will edit that section.

All in all it looks like implementing this is possible, but it will be hard to support multiple formats, and even supporting this only for markdown comes at the price of an additional dependency (kramdown).

**Comment by aransfor** [comment]
This would be awesome. The lack of section editing is on of the biggest downsides of gollum

**Comment by SkyCrawl** [comment]
It appears I don't see how this is so difficult? Look at how wikipedia does it:
`http://en.wikipedia.org/w/index.php?title=Information_technology&action=edit&section=1`

It references sections by integers, which is the best way :). Why should we ever concern ourselves with what Gollum generates in a section's `<a>` tag?

It's surely not that difficult to write a mechanism that captures source markup for a certain pair of consecutive `<h1>` tags (for each format - using regexes) and minds false positives in some cases?

Once you have the source section, you can easily split the page into "the up-context", "the source" and "the down-context". Even live preview should be (not sure since I didn't program it) easily editable to:
* display both contexts, if configured to do so by the user,
* define a canvas between contexts for "the source" and separate it visually,
* dynamically replace the content of the above canvas with the rendered content of the section.

Live-preview editor, as I take it, sends direct content of the left pane to be rendered and places it in a predefined canvas in the right pane. How is that different from the functionality I just mentioned? You only need to tweak the circumstances a bit :).

**Comment by dometto** [comment]
The difficulty is in mapping bits of the raw source (which needs to be in the edit field) to bits of the rendered html content. We can go look for h1's in the html output, but that tells us nothing about which lines of the source correspond to it. AFAIK the live preview does not implement anything like this at all (it doesn't hide/display specific bits of the source, does it?).

Apart from actually parsing the source into sections, gollum is perfectly ready to implement this - -  splitting the source into before and after bits and putting that in the editor form is the easy bit. :)

And yes, source maps is exactly what we need. But we need one for each markup kind (or decide to support this only for certain markups), and if they're implemented in Js, we have to make sure there are no significant differences in dialect between them and the ruby rendering gems.

**Comment by SkyCrawl** [comment]
> We can go look for h1's in the html output, but that tells us nothing about which lines of the source correspond to it.

Well, that's why we should match source markup that renders into `<h1>` (e.g. `^# ` in markdown) and capture the section that corresponds to a given section number? Then we split the page like @roccomoretti and me mentioned.

> The difficulty is in mapping bits of the raw source (which needs to be in the edit field) to bits of the rendered html content.

Not if we do it like suggested? Except, of course, some frustrating markup like footnotes, which may be defined somewhere else completely and only be referenced from the section :(... Despite there might be some workarounds for them too, maybe we don't need to care too much?

> AFAIK the live preview does not implement anything like this at all (it doesn't hide/display specific bits of the source, does it?).

What did you mean with `hide/display specific bits of the source, does it?`?
As far as I can tell, I suggested something else - the left pane (source markup) will only contain source markup for the given section. That markup will be sent to be rendered in Gollum and the response will be received. Then, the live-preview editor will only replace the content of a given canvas with the content of the rendered markup. THE MAGIC is to place this canvas like this (pseudocode):
```
<div id="live-preview-right-pane">
     <div id="up-context">
          This is a fixed HTML of rendered source markup, UP TO THE EDITED SECTION BEGINNING.
     </div>
     <div id="canvas">
          This is where the edited section will render.
     </div>
     <div id="down-context">
          This is a fixed HTML of rendered source markup, FROM THE END OF THE EDITED SECTION.
     </div>
</div>
```

Only the "canvas" will be dynamic and the rest of the preview fixed. It's also very easy to show/hide both the contexts - a little Javascript and 2 checkboxes.

**Comment by dometto** [comment]
@SkyCrawl, I think it's a little more complex than matching '# Section title' and analogoes for other markups. Consider:

```
# Section title
~~~~
# Section title
~~~~
```
Now we have to know not to match #'s in code blocks. In principle there is no end to such exceptions, so we'd have to almost implement a full parser for each markup that supports this feature to implement this well. :-(

Of course it's possible to just implement it with the simple match and hope it doesn't break too often, but I would much prefer something pretty with source maps. :)

**Comment by dometto** [comment]
Ehm, apparently github can't render a code block in a code block. I meant to give the example of someone who has a markdown heading within a triple backtick code block. :)

**Comment by SkyCrawl** [comment]
> I think it's a little more complex than matching '# Section title' and analogoes for other markups.

Hmm, right... Sorry :).

> Now we have to know not to match #'s in code blocks. In principle there is no end to such exceptions, so we'd have to almost implement a full parser for each markup that supports this feature to implement this well. :-(

Not necessarily. I think I can see a clever way to get around the issue of "infernal regexing", but for it to work, Gollum needs to remember "our own simple specific source maps", which means... request sessions :).

**Comment by SkyCrawl** [comment]
> Ehm, apparently github can't render a code block in a code block.

Try the following, WITH the indentation:

```
     ```
        //
     ```
```

**Comment by SkyCrawl** [comment]
### Okay, how about this (gonna speak with pseudocode again):
---
First and foremost, the request:
`localhost:4567/edit?page=path-to-page&section=1`

Then how the server will handle it:
```
P = get_page(param(path-to-page))
RS = raw_source(P)
# I and J are character offsets where the section begins/ends
I, J = get_section(RS, param(section))
# UC is "up-context", S is "section", DC is "down-context"
# S doesn't contain the markup where the "header" is defined? Only the section's content?
UC, S, DC = split(RS, I, J)
serve_page(render(UC), S, render(DC), I, J)     # I think I and J will be useful again later when saving the new content
```

And finally, the most interesting part: _get_section(RS, SN)_. For simplicity, I will assume single-process filter chain and markdown only.

1. Filter: expand macros.
2. Filter: tokenize headers:
    ```
    matched_index = 0
    foreach regex_matched(input, "^#(.*)$") do
        T = gen_token(matched_index)
        replace_line_with("#" + T + $1)
        matched_index++
    input -> tokenized
    ```
    Tokens need to be a special "syntax" (e.g. `<div id="12345" class="gollum-we-want-source-maaaaaps-gollum"/>`) that:
    * has a unique identifier, directly mapped to each "header syntax candidate",
    * is NOT parsed out by the renderer (it is ignored),
    * doesn't conflict with the syntax of any markup,
    * is sure not to be accidentally produced by users.

3. Filter: render. Simply render the result of the second filter.
4. Filter: map:
```
doc = Nokogiri.parse(input)
section_token = doc.css_search(" > h1:nth-of-type(SN) > .gollum-we-want-source-maaaaaps-gollum")
next_section_token = doc.css_search(" > h1:nth-of-type(SN+1) > .gollum-we-want-source-maaaaaps-gollum")
I = tokenized.index_of("^#" + section_token)
J = tokenized.index_of("^#" + next_section_token)
```

---

**SUMMARY:**
It may look complicated but it's actually rather simple. The worst part lies, in my opinion, somewhere else:
* the behaviour of the gem renderers,
* the markups - some of them may define some strange "blocks" which could potentially span across headers (sections) and cause rendering of our single section to fail.

**Comment by dometto** [comment]
@SkyCrawl, interesting idea. But again I wonder if this will really work out. Consider how the following piece of markdown is rendered (e.g. by `kramdown` on the CLI):
```
# <div id="12345" class="gollum-we-want-source-maaaaaps-gollum"/> My Header
```
Output:
```
<h1 id="div-id12345-classgollum-we-want-source-maaaaaps-gollum-my-header">&lt;div id="12345" class="gollum-we-want-source-maaaaaps-gollum"/&gt; My Header</h1>
```
So simply searching for the `gollum-we-want-source-maaaaaps-gollum` class will not work in this case (and there is no way to be sure it will work across different markups). :-(

I'm still hoping a good JS markdown lib will support source maps out of the box..

**Comment by SkyCrawl** [comment]
> So simply searching for the gollum-we-want-source-maaaaaps-gollum class will not work in this case

Ok, kramdown is a problem... By default, Gollum uses redcarpet though? :)

> (and there is no way to be sure it will work across different markups). :-(

Yes, I mentioned that. It needs to be tested... I don't think it would that difficult to find something that works across markups though - especially if it's just text, no special characters :). Before discarding the idea, we could even try escaping the tag: `&lt;div id="12345" class="gollum-we-want-source-maaaaaps-gollum"/&gt;`.

> I'm still hoping a good JS markdown lib will support source maps out of the box..

Even if, there are still many other markups to solve? :)
Anyway, it will probably take a VERY long time until all the renderers support source maps...

**Comment by dometto** [comment]
> By default, Gollum uses redcarpet though? :)

It used to be `github-markdown`, but we've actually switched to kramdown as the only markup gem dependency. But that shouldn't matter: gollum has always used whatever markdown renderer the user has installed (as long as it's recognized by `github-markup`).

>  I'd actually prefer something non-HTML as well :)

What about just making the tags something like `GOLLUM_SOURCE_MAP[n]`, with n the line number in the raw source where the header was defined, and then go look for `<h1>`'s whose content matches `/^GOLLUM_SOURCE_MAP[(\d+)](\s*)(\S+)/`?

**Comment by SkyCrawl** [comment]
I think `[n]` might sometimes be used as a footnote or citation ref? The syntax should be markup-resistant as well :). But in general, it's certainly an improvement to the HTML syntax I proposed :+1: .

**Comment by SkyCrawl** [comment]
How about just something like `gollumWONDERFULLYevasiveSOURCEmapsN` where the last N is a number?

**Comment by dometto** [comment]
> How about just something like gollumWONDERFULLYevasiveSOURCEmapsN where the last N is a number?

Will break for `#1` (the 1 will be interpreted as a digit in the line number). So we need a terminating string.. but yes, something like `GOLLUMSRCMP0GOLLUMSRCMP`. :)

**Comment by SkyCrawl** [comment]
Yep, looks like we have a winner :).

**Comment by dometto** [comment]
I like the approach. :+1: Sorry, I first thought it was dependent on an html implementation. But this should be very doable. I think the functionality can be implemented in a self-contained PR, although actually adding the edit section links might require some changes to existing filters.

**Comment by SkyCrawl** [comment]
Yes. There will be other various changes to do, including CSS and Gollum CAN support this at some form at the moment but I still think it would be better to wait for all the changes I'm making... They should make a lot of it simpler and especially the CSS :). On the other hand, god knows when I'll finish with all that unless there would be more of us to have a go at it.

**Comment by SkyCrawl** [comment]
Anyway, this discussion has become rather long... Shall I assemble the pieces and create a new issue?

**Comment by dometto** [comment]
> Shall I assemble the pieces and create a new issue?

Yes please. :-) I'll close this then. Practically I think the best place to start would be a simple PR implementing the `#get_section` method in gollum-lib, with the required tests. I can have a go at that later this week, unless you would prefer to. The implementation in the front end can always come at a later time.

**Comment by SkyCrawl** [comment]
It's alright, have a go at it yourself :). I have too much other work on my hands, unfortunately :(.

BTW, could you please create a `css-to-sass` branch for me in `gollum-lib`? That way, I could make some non-functioning PRs when I feel the time is right and you could examine the new designs I'm implementing, despite the coce would still be a mess at some places.

**Comment by dometto** [comment]
> could you please create a css-to-sass branch for me in gollum-lib?

Done! Thanks for all the thought and time you're putting into gollum.
